### ISSUE TITLE
Create a new developer guide

### ISSUE THREAD

**Issue by mmuejde** [scikit-activeml/scikit-activeml#186]
Please checkout the draft version

> branch name: @mmuejde/developerguide

**Comment by mmuejde** [comment]
Delete 'audience ....'

**Comment by mmuejde** [comment]
## Add 
- requirements.txt
- requirements_ext.txt

**Comment by christoph14** [comment]
Guide for pool-based query strategies:

- All query strategies are stored in a file skactiveml/pool/_query_strategy.py

- Every class inherits from SingleAnnotatorPoolBasedQueryStrategy

- The class must implement the __init__ finction for initialization and a 'query' function

- For typical class parameters we use standard names:
  - prior (Prior probabilities for the distribution of probabilistic strategies)
  - random_state (number or np.random.RandomState like sklearn)
  - method (string for classes that implement multiple methods)
  - cost_matrix (Cost matrix defining the cost of predicting instances wrong)

- Typical parameters of the query function are:
  - X_cand (Set of candidate instances, inherited from SingleAnnotatorPoolBasedQueryStrategy)
  - clf (The classifier used by the strategy)
  - X (Set of labeled and unlabeled instances)
  - y ((unknown) labels of X)
  - sample_weight (Weights of training samples in `X`)
  - sample_weight_cand (Weights of samples in `X_cand`)
  - batch_size (Number of instances for batch querying, inherited from SingleAnnotatorPoolBasedQueryStrategy)
  - return_utilities (inherited from SingleAnnotatorPoolBasedQueryStrategy)

- The query function returns:
  - query_indices (indices of the best instances)
  - utilities (utilities of all candidate instances, only if return_utilities is True)

- General advice for the 'query' code:
  - use self._validate_data function (is implemented in superclass)
  - check the input X and y only once
  - fit clf if it is not yet fitted (may use fit_if_not_fitted form utils)
  - calculate utilities (in an extra function)
  - use simple_batch function from utils for return value

- All query strategies are tested by a general unittest (test_pool.py):
  - Querying of every method is tested with standard configurations with 0, 1, and 5 initial labels.
  - For every class ExampleQueryStrategy that inherits from SingleAnnotPoolBasedQueryStrategy (stored in _example.py), it is automatically tested if there exists a file test/test_example.py. It is necessary that both filenames are the same. Moreover, the test class must be called TestExampleQueryStrategy(unittest.TestCase)
  - Every parameter in __init__() will be tested if it is written the same as a class variable.
  - Every parameter arg in __init__() will be evaluated if there exists a method in the testclass TestExampleQueryStrategy that is called test_init_param_arg().
  - Every parameter arg in query() will be evaluated if there exists a method in the testclass TestExampleQueryStrategy that is called test_query_param_arg().
  - Standard parameters 'random_state', 'X_cand', 'batch_size' and 'return_utilities' are tested and do not have to be tested in the specific tests.

**Comment by tpham93** [comment]
Guide for stream-based query strategies:
- All query strategies are stored in a file skactivml/stream/*.py
- Every query strategy inherits from `SingleAnnotatorStreamBasedQueryStrategy`
- Every query strategy has either an internal budget handling or an outsourced budget_manager
- The class must implement the following functions:
  - `__init__`: function for initialization
  - `query`: identify the instances whose labels to select
  - `update`: adapting the budget monitoring according to the queried labels
- For typical class parameters we use standard names:
  - `random_state`: integer that acts as random seed or `np.random.RandomState` like sklearn
  - `budget`: % of labels that the strategy is allowed to query
  - `budget_manager`: enforces the budget constraint
- Parameters of the query function are (similar to pool):
  - `X_cand` (Set of candidate instances, inherited from `SingleAnnotatorStreamBasedQueryStrategy`)
  - `clf` (The classifier used by the strategy)
  - `X` (Set of labeled and unlabeled instances)
  - `y` (labels of `X` (it may be set to `MISSING_LABEL` if `y` is unknown))
  - `sample_weight` (weights for each instance in `X` or `None` if all are equally weighted)
  - `return_utilities` (inherited from `SingleAnnotatorStreamBasedQueryStrategy`)
- The `query` function returns:
  - `queried_indices` (indices of the best instances from `X_Cand`)
  - `utilities` (utilities of all candidate instances, only if `return_utilities` is `True`)
- General advice for `query` code:
  - the query function must not change the internal state of the query strategy (`budget` and `random_state` included) to allow for assessing multiple instances with the same state. Update the the internal state in the `update()` function
  - use `self._validate_data` function (is implemented in superclass)
  - check the input `X` and `y` only once
  - fit classifier if it is not yet fitted (may use `fit_if_not_fitted` from `utils`)
- Typical parameters of the update function are:
  - `X_cand` (Set of candidate instances, inherited from `SingleAnnotatorStreamBasedQueryStrategy`)
  - `queried_indices` (typically the return value of `query`)
  - `budget_manager_param_dict`: provides additional parameters to the `update` function of the `budget_manager` (only include if a `budget_manager` is used)
- general advice for `update` code:
  - use `self._validate_data` in case the strategy is used without using the `query` method (if parameters need to be initialized before the update)
  - If a `budget_manager` is used forward the `update` call to the `budget_manager.update` method
- All stream query strategies are tested by a general unittest (stream/tests/test_stream.py):
  - For every class `ExampleQueryStrategy` that inherits from `SingleAnnotStreamBasedQueryStrategy` (stored in _example.py), it is automatically tested if there exists a file test/test_example.py. It is necessary that both filenames are the same. Moreover, the test class must be called `TestExampleQueryStrategy` and inherit from `unittest.TestCase`
  - Every parameter in `__init__()` will be tested if it is written the same as a class variable.
  - Every parameter arg in `__init__()` will be evaluated if there exists a method in the testclass `TestExampleQueryStrategy` that is called `test_init_param_arg()`.
  - Every parameter `arg` in `query()` will be evaluated if there exists a method in the testclass `TestExampleQueryStrategy` that is called `test_query_param_arg()`.

Guide for budget_manager for stream-based query strategies:
- All budget managers are stored in skactivml/stream/budget_manager/*.py
- The class must implement the following functions:
  - `__init__`: function for initialization
  - `query_by_utilities`: identify which instances to query based on the assessed utility
  - `update`: adapting the budget monitoring according to the queried labels
- The update function of the budget manager has the same functionality as the query strategy update
- For typical class parameters we use standard names:
  - `budget`: % of labels that the strategy is allowed to query
  - `random_state`: integer that acts as random seed or `np.random.RandomState` like sklearn
- Typical parameters of the `query_by_utilities` function are:
  - `utilities` (The `utilities` of `X_cand` calculated by the query strategy, inherited from `BudgetManager`)
- General advice for working with a `budget_manager`:
  - If a `budget manager` is used, the `_validate_data` of the `query strategy` needs to be adapted accordingly:
    - If only a `budget` is given use the default `budget_manager` with the given budget
    - If only a `budget_manager` is given use the `budget_manager`
    - If both are not given use the default `budget_manager` with the default `budget`
    - If both are given and the `budget` differs from `budget_manager.budget` throw an error
- All budget managers are tested by a general unittest (stream/budget_manager/tests/test_budget_manager.py):
  - For every class `ExampleBudgetManager` that inherits from `BudgetManager` (stored in _example.py), it is automatically tested if there exists a file test/test_example.py. It is necessary that both filenames are the same. Moreover, the test class must be called `TestExampleBudgetManager` and inheriting from `unittest.TestCase`
  - Every parameter in `__init__()` will be tested if it is written the same as a class variable.
  - Every parameter `arg` in `__init__()` will be evaluated if there exists a method in the testclass `TestExampleQueryStrategy` that is called `test_init_param_arg()`.
  - Every parameter `arg` in `query_by_utility()` will be evaluated if there exists a method in the testclass `TestExampleQueryStrategy` that is called `test_query_by_utility _param_arg()`.

**Comment by LukasLuehrs** [comment]
Guide for multi-annotator pool-based query strategies:

 - All query strategies are stored in a file skactiveml/pool/multi/_query_strategy.py

 - Every class inherits from `MultiAnnotatorPoolBasedQueryStrategy`

 - The class must implement the following functions:
    - `__init__`: function for initialization of hyperparameters
    - `query`: identify the instance annotator pairs whose labels to select

 - For typical class parameters we use standard names:
    - `random_state` (number or `np.random.RandomState` like sklearn)

 - Typical parameters of the query function are:
    - `X_cand` (Sequence of candidate instances to be queried, inherited from `MultiAnnotatorPoolBasedQueryStrategy`)
    - `A_cand` (Boolean mask further specifying which annotator can be queried for which candidate instance, inherited from `MultiAnnotatorPoolBasedQueryStrategy`)
    - `clf` (The classifier used by the strategy)
    - `X` (Sequence of labeled and unlabeled instances)
    - `y` ((unknown) Labels of X for each annotator)
    - `sample_weight` (Weights of the prediction of a sample from an annotator (used for predictions of labels))
    - `A_perf` (Performance of an annotators for a given sample, usually the accuracy (used for estimating the best annotator to query for a given candidate sample))
    - `batch_size` (Number of instances for batch querying, inherited from `MultiAnnotatorPoolBasedQueryStrategy`)
    - `return_utilities` (Inherited from `MultiAnnotatorPoolBasedQueryStrategy`)

 - The query function returns:
    - `query_indices` (Indices of the best candidate instance annotator pair)
    - `utilities` (Utilities of all candidate instances annotator pairs, only if `return_utilities` is True)

 - General advice for the query code:
    - use `self._validate_data` function (is implemented in superclass)
    - check the input `X` and `y` only once
    - fit classifier if it is not yet fitted (may use `fit_if_not_fitted` form utils)
    - if the strategy combines a single annotator query strategy with a performance estimate 
       - define an aggregation function
       - evaluate the performance for each annotator sample pair
       - use the `MultiAnnotWrapper`
    - if the strategy is a greedy method regarding the utilities 
       - calculate utilities (in an extra function)
       - use `simple_batch` function from utils for return value

**Comment by mherde** [comment]
**Guide for classifiers:**

- Standard classifier implementations are part of the subpackage `skactiveml.classifier` and classifiers learning from multiple annotators are implemented in its subpackage `skactiveml.classifier.multi`.

- Every class of a classifier inherits from `skactiveml.base.SkactivemlClassifier`.

- The class of a classifier must implement the `__init__` method for initialization, a `fit` method for training, and a `predict_proba` method predicting class membership probabilities for samples. A `predict` method is already implemented in the superclass by using the outputs of the `predict_proba` method. Additionally, a `score` method is implemented by the superclass to evaluate the accuracy of a fitted classifier.

- A commonly used subclass of `skactiveml.base.SkactivemlClassifier` is the `skactiveml.base.ClassFrequencyEstimator`, which requires an implementation of the method `predict_freq`, which can be interpreted as prior parameters of a Dirichlet distribution over the class membership probabilities of a sample.

- Typical `__init__` parameters are:
   
   - `classes`: Holds the label for each class. If `None`, the classes are determined during the fit.
   - `missing_label`: Value to represent a missing label.
   - `cost_matrix`: Cost matrix with `cost_matrix[i,j]` indicating cost of predicting class `classes[j]` for a sample of class `classes[i]`. Can be only set, if `classes` is not `None`.
   - `random_state`: Ensures reproducibility (cf. scikit-learn).
   - `class_prior`: A `skactiveml.base.ClassFrequencyEstimator` requires additionally  this parameter as prior observations of the class frequency estimates. 
 
- Required `fit` parameters are:

- Required `predict_proba`, `predict_freq`, and `predict` parameters are:

- Required `score` parameters are:

**Comment by mherde** [comment]
**Guide for annotator models:**

- Annotator models are marked by implementing the interface `skactiveml.base.AnnotMixing`. These models can estimate the performances of annotators for given samples.

- Every class of a classifier inherits from `skactiveml.base.SkactivemlClassifier`.

- The class of an annotator model must implement the `predict_annot_perf` method estimating the performances per sample of each annotator as proxies of the provided annotation's qualities.
 
- Required `predict_annot_perf` parameters are:
  - `X`: Is a matrix of feature values representing the samples. 

- `predict_annot_perf` method returns:
   - `P_annot`: The estimated performances per sample-annotator pair.

- General advice for the `predict_annot_perf` code:
   - Check parameter `X` regarding its shape.
   - Check that the annotator model has been fitted.
   - If no samples or class labels were provided during the previous call of the fit method, the maximum value of annotator performance should be outputted for each sample-annotator pair.
